### PR TITLE
[GUI] Avoid raw pointers for m_windowXMLRootElement

### PR DESCRIPTION
**Pull request by enen92**
## Description
As the title indicates, this avoids using raw pointers for one of the members of `GUIWindow`. I could not use the `default` destructor because `TiXmlElement` is an incomplete type (hence the empty destructor). However `CPlayerCoreConfig` seems to be able to do it just fine...

### CODE REVIEW

File: xbmc/guilib/GUIWindow.cpp [hunk 1]
```diff
@@ -123,7 +121,7 @@ bool CGUIWindow::Load(const std::string& strFileName, bool bContainsPath)
 bool CGUIWindow::LoadXML(const std::string &strPath, const std::string &strLowerPath)
 {
   // load window xml if we don't have it stored yet
-  if (!m_windowXMLRootElement)
+  if (m_windowXMLRootElement == nullptr)
```
Comment (ksooo) [inline xbmc/guilib/GUIWindow.cpp:124]: Why not stick with how it was - `if (!m_win...` ? `std::unique_ptr` has `bool` operator as far as i know.

File: xbmc/guilib/GUIWindow.h [hunk 2]
```diff
@@ -273,7 +273,8 @@ class CGUIWindow : public CGUIControlGroup, protected CCriticalSection
   CGUIAction m_loadActions;
   CGUIAction m_unloadActions;
 
-  TiXmlElement* m_windowXMLRootElement;
+  ///< \brief window root xml definition after resolving any skin includes. Stored to avoid parsing the XML every time the window is loaded.
```
Comment (ksooo) [inline xbmc/guilib/GUIWindow.h:276]: This is an ugly far too long line. ;-)

File: xbmc/guilib/GUIWindow.cpp [hunk 3]
```diff
@@ -792,8 +791,7 @@ void CGUIWindow::FreeResources(bool forceUnload /*= false */)
   if (m_loadType == LOAD_EVERY_TIME || forceUnload) ClearAll();
   if (forceUnload)
   {
-    delete m_windowXMLRootElement;
-    m_windowXMLRootElement = nullptr;
+    m_windowXMLRootElement.release();
```
Comment (ksooo) [inline xbmc/guilib/GUIWindow.cpp:794]: Better use `reset()`, which also does what you want but returns `void`, in difference to `release`which returns the raw pointer to the released object. But you do not need that pointer, right?

File: xbmc/guilib/GUIWindow.cpp [hunk 4]
```diff
@@ -145,21 +143,22 @@ bool CGUIWindow::LoadXML(const std::string &strPath, const std::string &strLower
...
+std::unique_ptr<TiXmlElement> CGUIWindow::Prepare(const std::unique_ptr<TiXmlElement>& rootElement)
 {
-  if (!pRootElement)
+  if (!rootElement)
     return nullptr;
 
   // clone the root element as we will manipulate it
-  auto preparedRoot = std::unique_ptr<TiXmlElement>(static_cast<TiXmlElement*>(pRootElement->Clone()));
+  auto preparedRoot =
+      std::unique_ptr<TiXmlElement>(static_cast<TiXmlElement*>(rootElement->Clone()));
```
Comment (ksooo) [inline xbmc/guilib/GUIWindow.cpp:161]: Use `std::make_unique`.

**Review by ksooo — APPROVED** [review]
LGTM

File: xbmc/guilib/GUIWindow.cpp [hunk 5]
```diff
@@ -145,21 +143,22 @@ bool CGUIWindow::LoadXML(const std::string &strPath, const std::string &strLower
...
+std::unique_ptr<TiXmlElement> CGUIWindow::Prepare(const std::unique_ptr<TiXmlElement>& rootElement)
 {
-  if (!pRootElement)
+  if (!rootElement)
     return nullptr;
 
   // clone the root element as we will manipulate it
-  auto preparedRoot = std::unique_ptr<TiXmlElement>(static_cast<TiXmlElement*>(pRootElement->Clone()));
+  auto preparedRoot =
+      std::make_unique<TiXmlElement>(*static_cast<TiXmlElement*>(rootElement->Clone()));
```
Comment (notspiff) [inline xbmc/guilib/GUIWindow.cpp:161]: why call Clone if TiXmlElement has a copy constructor?
```c++
auto preparedRoot = std::make_unique<TiXmlElement>(*RootElement))
```
should work and avoid the leak.

**Review by garbear — APPROVED** [review]
PR is fine in its current form and is certainly an improvement. You might be able to go further though - see the pattern where a pointer is passed and we always check if it's null? I suspect all these nullptr checks can be skipped if you just pass everything around by const reference:

```c++
std::unique_ptr<TiXmlElement> CGUIWindow::Prepare(const TiXmlElement& rootElement)
```

But maybe this assumption can't be made and we intend to pass empty pointers in some cases.

**Comment by enen92** [comment]
Jenkins build this please

**Comment by enen92** [comment]
Jenkins build and merge

**Pull request merged**